src/build/__main__.py

```diff
@@ -24,6 +24,24 @@
 __all__ = ['build', 'main', 'main_parser']
 
 
+_STYLES = {
+    'red': '\33[91m',
+    'green': '\33[92m',
+    'yellow': '\33[93m',
+    'bold': '\33[1m',
+    'dim': '\33[2m',
+    'underline': '\33[4m',
+    'reset': '\33[0m',
+}
+
+
+def _print(message: str) -> None:
+    color = sys.stdout.isatty() and 'NO_COLOR' not in os.environ
+    for name, code in _STYLES.items():
+        message = message.replace(f'[{name}]', code if color else '')
```

Comment:
Why not just use string formatting instead of ad hoc replaces?  Something like:

```py
def _print(message: str) -> None:  # pragma: no cover
    color = sys.stdout.isatty() and 'NO_COLOR' not in os.environ
    if color:
        message = message.format(**_STYLES)
    else:
        message = message.format(**{k: '' for k in _STYLES})
    print(message)

[...]

_print(f'{{yellow}}WARNING{{reset}} {message}')
```

Reply:
The performance difference is not relevant, and I felt this had slightly cleaner code and considerably better readability in the messages -- the color formatting is clearly distinct from f-string variables and only takes one delimiter character on each side.

```python
import os
import sys


_STYLES = {
    'a': '1',
    'b': '2',
    'c': '3',
    'd': '4',
    'e': '5',
}


def _print_replace(message: str) -> None:
    color = sys.stdout.isatty() and 'NO_COLOR' not in os.environ
    for name, code in _STYLES.items():
        message = message.replace(f'[{name}]', code if color else '')
    #print(message)


def _print_format(message: str) -> None:  # pragma: no cover
    color = sys.stdout.isatty() and 'NO_COLOR' not in os.environ
    if color:
        message = message.format(**_STYLES)
    else:
        message = message.format(**{k: '' for k in _STYLES})
    #print(message)
```

```
$ python -m timeit -s 'import test_print' -- 'test_print._print_replace("[a][b][c][d][e]")'
100000 loops, best of 5: 2.35 usec per loop
$ python -m timeit -s 'import test_print' -- 'test_print._print_format("{{a}}{{b}}{{c}}{{d}}{{e}}")'
200000 loops, best of 5: 1.93 usec per loop
```

Reply:
I don't think the performance difference is important either - I just felt it's the obvious way to do string substitution. Let's keep the squares then if you think readability would suffer.

Reply:
This looks a lot like rich's mini language - why not exactly imitate that? `[yellow]<stuff>[/yellow]`. You could loop over over styles and add `/` versions that reset. Just a thought, instead of inventing something new just for this. _Very_ minor though.

Reply:
Also what happens if the string `[yellow]` happens to be part of the traceback, since tracebacks are passed through this? Doing fstring replacement wouldn't have this problem.

Reply:
More complexity in the implementation :stuck_out_tongue:. This is private anyway, so I just went with the simple thing that would fit our needs.

Reply:
> Also what happens if the string `[yellow]` happens to be part of the traceback, since tracebacks are passed through this? Doing fstring replacement wouldn't have this problem.

Right, we should not pass them through.

Reply:
Personally, in plumbum.colorlib, I use `print(yellow | "str") and `yellow` wraps the string it's or'd with with color tags.

Reply:
Right, but that introduces a bit more complexity than what we need. I opened #335.